src/types/closure.md

```diff
@@ -178,3 +294,258 @@ Because captures are often by reference, the following general rules arise:
+            * Return (Place, Mode)
+        * Else if place contains a deref:
+            * Let Place1 = Place truncated just before the deref
+            * Return (Place1, ByValue)
+        * Else:
+            * Return (Place, ByValue)
+    * `ref_xform(Place, Mode) -> (Place, Mode)` (for ref closures)
+        * "If taking ownership of data, only move data from enclosing stack frame."
+        * Generate C' by mapping each (Mode, Place) in C
+            * If Mode is ByValue and place contains a deref:
+                * Let Place1 = Place truncated just before the deref
+                * Return (Place1, ByValue)
+            * Else:
+                * Return (Place, Mode)
+    * `ref_opt(Place, Mode) -> (Place, Mode)`
+        * "Optimization: borrow the ref, not data owned by ref."
+        * Disjoint capture over immutable reference doesn't add too much value because the fields can either be borrowed immutably or copied.
+        * Edge case: Field that is accessed via the referece lives longer than the reference.
+            * Resolution: Only consider the last Deref
+        * If Place is (Base, Projections), where Projections is a list of size N.
+            * For all `i, 0 <= i < N`, Projections[i] != Deref
+                * Return (Place, Mode)
+            * If `l, 0 <= l < N` is the last/rightmost Deref Projection i.e. for any `i, l < i < N` Projection[i] != Deref,
+              and `Place.type_before_projection(l) = ty::Ref(.., Mutability::Not)`
+                * Let Place1 = (Base, Projections[0..=l])
+                * Return (Place1, Ref)
+
+## Key examples
+
+### box-mut
+
+```rust
+struct Foo { x: i32 }
+
+fn box_mut() {
+    let mut s = Foo { x: 0 } ;
+    
+    let px = &mut s;
+    let bx = Box::new(px);
+    
+    
+    let c = move || bx.x += 10;
+    // Mutable reference to this place:
+    //   (*(*bx)).x
+    //    ^ ^
+    //    | a Box
+    //    a &mut
+}
+```
+
+<!-- ignore: Omit error about unterminated string literal when representing c_prime -->
+```ignore
+Closure mode = move
+C = {
+    (ref mut, (*(*bx)).x)
+}
+C' = C
+```
+
+Output is the same: `C' = C`
+
+### Packed-field-ref-and-move
+
+When you have a closure that both references a packed field (which is unsafe) and moves from it (which is safe) we capture the entire struct, rather than just moving the field. This is to aid in predictability, so that removing the move doesn't make the closure become unsafe:
+
+```rust
+#[repr(packed)]
+struct Packed { x: String }
+
+# fn use_ref<T>(_: &T) {}
+# fn move_value<T>(_: T) {}
+
+fn main() {
+    let packed = Packed { x: String::new() };
+
+    let c = || {
+        use_ref(&packed.x);
+        move_value(packed.x);
+    };
+
+    c();
+}
+```
+
+<!-- ignore: Omit error about unterminated string literal when representing c_prime -->
+```ignore
+Closure mode = ref
+C = {
+    (ref mut, packed)
+}
+C' = C
+```
+
+### Optimization-Edge-Case
+```edition2021
+struct Int(i32);
```

Comment:
@ehuss  @nikomatsakis should I just make this ignore since tests seem to not allow edition 2021 since it's unstable?

Reply:
By the time this lands, they will, I believe